My enum methods

jelduran opened the pull request:
This is a pull request for Enumerable Methods Project in Microverse Main Curriculum, Ruby Basic Course.

It is about an own version of some of the main methods for Array and Hash Objects in Ruby language.
These are the developed methods:
my_each
my_each_with_index
my_select
my_all?
my_any?
my_none?
my_count
my_map
my_inject
multiply_els

jelduran committed:
Linter:

Linter Rubocop added to workflow.

Commit:
Created main.rb:

Module Enumerable created in the file main.rb

Commit:
gitignore:

Added gitignore to the repository.

jelduran committed:
Ignored:

Folder .idea ignored and deleted

jelduran committed:
Method my_each:

Method my_each defined.

Commit:
my_each variable:

Created keys variable to store hash keys if object is a hash.

jelduran committed:
my_each condition:

Check if method my_each gets a block using block_given? method.

Commit:
my_each statement:

This instruction is to iterate through the object if it is an Array, using the length of the object as an argument for times method, executing a block with index as argument and using yield method for a next block passing the item as an argument.

jelduran committed:
my_each statement:

This instruction is to iterate through the object if it is an Hash, using the length of the object as an argument for times method, executing a block with index as argument and using yield method for a next block, using the keys variable to get the key and the value of this Hash item and pass it like an Array object.

Commit:
my_each return:

my_each method return its caller object.

jelduran committed:
my_each return:

If method doesn't get a block then it returns an Enumerator object made of caller object.

Commit:
my_each_with_index method:

This method is basically a clone of my_each method the only difference is it also send the value of the index to the next block.

jelduran committed:
Method my_select:

Method my_select defined.

Commit:
my_select variable:

A new enumerable object is created if caller object is a Hash using a ternary operator assignment.

jelduran committed:
my_select condition:

Check if method my_select gets a block using block_given? method.

Commit:
my_select statement:

This instruction is using my_each method to go through the object, sending the item as an argurment and applying a block of instructions, pushing the return of the block in the new_enumerable object created at the beginning, all this if the caller object is an Array.

Commit:
my_select statement:

This instruction is to iterate through the object if it is an Hash, using the length of the object as an argument for times method, executing a block with index as argument and using yield method for a next block, using the keys variable to get the key and the value of this Hash item and pass it like an Array object..

jelduran committed:
my_select statement:

This instruction is using my_each method to go through the object, sending the item (array with a key,value pair) as an argurment and applying a block of instructions, pushing the return of the block in the new_enumerable object created at the beginning, all this if the caller object is an Hash.

jelduran committed:
Merge branch 'my-enum-methods' of github.com:jelduran/Advanced-Building-Blocks---Enumerables into my-enum-methods

Commit:
my_select return:

my_select method returns the new_enumerable object.

jelduran committed:
my_select return:

If method doesn't get a block then it returns an Enumerator object made of caller object.

Commit:
Method my_all?:

Method my_all? defined with the optional parameter pattern to manage regular expressions.

jelduran committed:
my_all? condition:

Check if method my_all? gets a block using block_given? method.

jelduran committed:
my_select condition:

Check if method my_all? gets a block using block_given? method.

Commit:
Merge branch 'my-enum-methods' of github.com:jelduran/Advanced-Building-Blocks---Enumerables into my-enum-methods

Commit:
my_all? statement:

This instruction is using my_each method to go through the object, sending the item as an argurment and applying a block of instructions, returning a boolean according to the result of evaluating the item in the next block, all this if the caller object is an Array.

jelduran committed:
my_select statement:

This instruction is using my_each method to go through the object, sending the item as an argurment and applying a block of instructions, returning a boolean according to the result of evaluating the item (array with key,value pair) in the next block, all this if the caller object is an Hash.

jelduran committed:
my_all? statement:

This instruction is using my_each method to go through the object, sending the item as an argurment and applying a block of instructions, returning a boolean according to the result of evaluating the item (array with key,value pair) in the next block, all this if the caller object is an Hash.

Commit:
Merge branch 'my-enum-methods' of github.com:jelduran/Advanced-Building-Blocks---Enumerables into my-enum-methods

Commit:
my_all? return:

my_all? method return true if none of the previous statement return false.

jelduran committed:
my_all? recursive call:

If no block is given and if a pattern is received as an argument then the method itself is called recursively sending to it a block with a comparison statement between the item and the pattern argument.

Commit:
my_all? recursive call:

If no block is given and if a pattern is received as an argument then the method itself is called recursively sending to it a block with the item.

jelduran committed:
my_any? method:

This method is basically a clone of my_all? method the only difference is it is checking if any item in the collection accomplishes the condition in the given block if not it returns false.

Commit:
my_none? method:

This method is basically a clone of my_any? method the only difference is it is checking if none item in the collection accomplishes the condition in the given block if not it returns true.

jelduran committed:
Method my_count

Method my_count defined with the optional parameter search to manage how many times is a given iten in the caller object.

Commit:
my_count variable:

Created counter variable to store how many items accomplish a condition in the given block.

jelduran committed:
my_count condition:

Check if method my_count gets a block using block_given? method.

Commit:
my_count statement:

This instruction is using my_each method to go through the object, sending the item as an argurment and applying a block of instructions, adding 1 to counter variable created at the beginning, all this if the caller object is an Array.

jelduran committed:
my_count statement:

This instruction is using my_each method to go through the object, sending the item as an argurment and applying a block of instructions, adding 1 to counter variable created at the beginning, all this if the caller object is an Hash.

Commit:
my_count return:

my_count method returns the counter variable.

jelduran committed:
my_count recursive call:

If no block is given and if a search criteria is received as an argument then the method itself is called recursively sending to it a block with a comparison statement between the item and the search argument.

Commit:
my_count return:

The method returns the length of the caller object.

jelduran committed:
Method my_map:

Method my_map is defined.

Commit:
my_map variable:

Created map variable to store the result of the given block for every item in the caller object.

jelduran committed:
my_map condition:

Check if method my_map gets a block using block_given? method.

Commit:
my_map statement:

This instruction is using my_each method to go through the object, sending the item as an argurment and applying a block of instructions, pushing the result to map variable created at the beginning, all this if the caller object is an Array.

jelduran committed:
my_map statement:

This instruction is using my_each method to go through the object, sending the item (array of key,value pair) as an argurment and applying a block of instructions, pushing the result to map variable created at the beginning, all this if the caller object is an Hash

Commit:
my_map return:

The method returns an array of mapped items.

jelduran committed:
my_map return:

If method doesn't get a block then it returns an Enumerator object made of caller object.

Commit:
Method my_inject:

Method my_inject is defined.

jelduran committed:
my_inject statement:

This instruction is using my_each method to go through the object, sending the item as an argurment and applying a block of instructions, accumulating the result to memo argument or assignning memo firs item value if memo is null, all this if the caller object is an Array.

Commit:
my_inject statement:

This instruction is using my_each method to go through the object, sending the item (array of key, value pair) as an argurment and applying a block of instructions, accumulating the result to memo argument or assignning memo firs item value if memo is null, all this if the caller object is an Hash.

jelduran committed:
my_inject return:

Method returns the final value of memo argument.

Commit:
Method multiply_els:

Method multiply_els is defined.

jelduran committed:
multiply_els statement:

Method my_inject is invoked to caller object giving &:* (multiplication) as a Proc object argument.

Commit:
my_map change:

Method my_map has been changed to get a Proc object argument instead of block object argument and call method has been used instead of yield method.

jelduran committed:
my_map change:

A new condition has been added to check if a block is given with the same functionality, the priority is for the Proc object argument before than a implicit block.

Commit:
Readme File modified:

Added the general information of the project.

Commit:
Readme File modified:

Added contact information.

Commit:
Readme File modified:

Added contact information.

jelduran committed:
Rubocop 1:

Fixed coding problems about bad spacing, perceived complexity, bad use of literal assignments and bad use of triple equal operator.

jelduran committed:
Rubocop 2:

Fixed coding problems about cyclomatic complexity.

jelduran committed:
Readme File modified:

Screenshot and Live Version link added.

jelduran committed:
Git ignore:

Files added to .gitignore.

jelduran committed:
Git ignore:

Files removed from repo.

kitpao commented on README.md:
```diff
@@ -1,2 +1,79 @@
-Own version of Ruby’s enumerable methods.
+By Carlos Robles & John Durán.
+
+This project we used Ruby to build it.
```
- [ ] **OPTIONAL** For grammar purposes, please use 'We used Ruby to build this project' or 'This project was built with Ruby'

kitpao commented on README.md:
```diff
@@ -1,2 +1,79 @@
+To get a local copy up and running follow these simple example steps.
+
+Open the terminal command line window
+git clone https://github.com/jelduran/Advanced-Building-Blocks---Enumerables.git
```
- [ ] **OPTIONAL** For formatting, you can use ( ` ) backquotes to indicate a line of code
```
This is written with 3 backquotes at the beginning and 3 backquotes at the end
```
This is written with `one backquote at the beginning and 1 at the end`
![image](https://user-images.githubusercontent.com/53256139/90822222-b931bc80-e2f9-11ea-8117-a28f9ae6e885.png)

kitpao commented on .gitignore:
```diff
@@ -0,0 +1,4 @@
+.idea
+.rubocop.yml
+Gemfile
+Gemfile.lock
```
- [ ] Please don't ignore Gemfile and Gemfile.lock, they contain the gems and the versions you are using so that other persons can make use of your repo.

kitpao reviewed (changes requested):
# 👋 Hello @jelduran  and @carlos-ssh 

🌟 Good job setting your repo correctly with the new Git Flow. ✏️ I left your feedback below and on the inline comments, please check them 🔽 

**_Note:_** All the **[OPTIONAL]** are **not required** in order to approve your project but are personal suggestions to improve your coding experience

- [ ] Your commits are descriptive 👍 but for next time, try to follow [this guideline indicated in Pathwright for best practices](https://www.freecodecamp.org/news/level-you-up-to-awesome-commit-messages-a85558cb90e8/) like starting your commits with a verb in the imperative tense

## Methods:
_In this section, you will find the differences between the original methods and your methods with some hints to fix them._
- #my_each
  - [ ] `p [1,2,3].each{ |x| x }` should return `[1, 2, 3]`, your code returns `#<Enumerator: [1, 2, 3]:my_each>` **Hint:** That is because it is returning the last line in your code `to_enum(:my_each)`, you should return what you have in your condition. The same happens in lines 9 and 19
  - [ ] **OPTIONAL** To avoid problems with the use of returns inside ifs, you can also use (this can apply to all your methods that return an enumerable when there is no block):
    ```
    def my_each
      return to_enum(:my_each) unless block_given? #<-- Proposed line
      
      # The rest of your logic without ifs here
      self #<-- value returned automatically
    end
    ```
  - [ ] `(1..3).my_each{ |x| print x }` should print `123`, your method prints nothing. **Hint:** Your my_each doesn't yield input when it is a range, so it is not grabbing the info.
  - [ ] **OPTIONAL** Why don't you convert the input in an array since the beginning? so you don't have to be checking if its a hash or an array or a range. 
- #my_each_with_index
  - [ ] The failed tests are due to the exact same issues mentioned in #my_each 
- #my_select
  - [ ] It has the same issue of not accepting a range
- #my_all?
  - [ ] `[3, 3, 3].my_all?(3)` should return true, your method returns false. **Hint:** That is because you forgot the case were the "pattern" is a number
- #my_any?
  - [ ] Same issue you had in #my_all
- #my_none?
  - [ ] Same issue you had in #my_all and #my_any
- #my_count
  - [ ] `(1..10).my_count` should return 10, but your method doesn't accept a range. **Hint:** you can convert the input into an array
  - [ ] `[1,2,3,3].my_count(3)` should return 2, but it returns 4 **Hint:** in the line where you use "if search", I think you wanted to return that line, but your code is returning the length
- #my_map
  - [ ] Your issues here are solved when you fix #my_each_with_index
  - [ ] **OPTIONAL** since you are not really using the index, for efficiency, you can simple use my_each
- #my_inject
  - [ ] `[1,2,3].my_inject(:+)` should return 6 (1+2+3), your method doesn't know what to do without a block Hint: Your programs need to use the symbol when it is given
- #multiply_els
  - [ ] `multiply_els [2, 4, 5]` should return 40, your method is undefined for an array **Hint:** take that method out of the module and pass the arguments manually

### ♻️ Status: Your project requires some additional changes ♻️
➡️ Please request another review once you are ready 📝 
___
##### 👉 If you have any doubts you can contact me in slack as @kitpao

🔗 __[Github](https://github.com/kitpao) | [Twitter](https://twitter.com/Kitpao1) | [LinkedIn](https://www.linkedin.com/in/kitzia-paola-vidal/) | [Medium](https://medium.com/@kitpao91)__

jelduran committed:
Use guard clause:

All methods logic has been changed using a guard clause at first instead of returning at the end, as TSE recommended.

jelduran committed:
Convert to array:

my_each methods both were refactorized converting caller object to array and removing some unnecessary instructions.

jelduran committed:
Fix predicate methods:

Fix the logic in predicate methods that was wrong after change the guard clause in them.

jelduran committed:
Get symbol:

my_inject method is now working with symbol argument with no need to use & operator.

jelduran committed:
Set out multiply_els:

multiply_els is now out of Enumerable Module and it has also a parameter defined.

jelduran committed:
Fix readme:

Fixed some errors and formatting problems in readme file.

jelduran committed:
Add Gemfile:

Gemfiles removed from gitignore file.

jelduran committed:
Rubocop 3:

Gemfile validation problems fixed.

kitpao commented on main.rb:
```diff
@@ -0,0 +1,112 @@
+      case pattern.class.to_s
+      when 'Class' then my_any? { |item| item.is_a?(pattern) }
+      when 'Regexp' then my_any? { |item| item =~ pattern }
+      else my_all? { |item| item == pattern }
```
- [ ] test `[3, 1, 3].my_any?(3)` should be true, it returns false, your mistake is in these lines, I'm sure you can figure it out, it's some kind of a 'typo'

kitpao reviewed (changes requested):
# 👋 Hi again @jelduran and @carlos-ssh 

Great job guys! almost all your tests are passing. 🥳 

✏️ Please look at these remaining tests:

- [ ] You have an issue with #my_any, please check inline comment below 🔽 
- [ ] There is another issue in #my_none and it's the same as in #my_any. 
- [ ] For my_inject, nice! now it is accepting input without a block but doesn't know what to do with an initial value. `[1,2,3].my_inject(20, :+)` should return 26 (20+1+2+3). Remember it should accept ranges too.

You are almost there! You did nicely fixing almost everything

### ♻️ Status: Your project requires some additional changes ♻️
➡️ Please request another review once you are ready 📝 
___
##### 👉 If you have any doubts you can contact me in slack as @kitpao

🔗 __[Github](https://github.com/kitpao) | [Twitter](https://twitter.com/Kitpao1) | [LinkedIn](https://www.linkedin.com/in/kitzia-paola-vidal/) | [Medium](https://medium.com/@kitpao91)__

jelduran committed:
Fix typos:

Fixed typos in my_any? and my_none? methods.

jelduran committed:
Fix my_inject:

Fixed problem in my_inject method with memo argument when a sym argument is received.

jelduran committed:
Fix typos:

Fixed typo in my_none? method.

mikenath223 reviewed (approved):
### CONGRATULATIONS  👏👏👏

This project meets the requirements for approval.

**STATUS: APPROVED ✅**


Don't forget to fill the project completion form👍
Good luck with the next project.

HAPPY CODING👨‍💻

jelduran commented:
Thank you @kitpao and @mikenath223 for your reviews and great feedback!
